### PR TITLE
Move GHC session to a Shake rule

### PR DESCRIPTION
**Pull request by cocreature**
This is a spin off from the work for making module prefixing
work. There I wanted to be able to use the `GeneratePackageMap` rule
to get access to the modules in an LF package which requires to be in
an Action rather than in IO.

This resulted in a much larger refactoring than I expected, so here is
a list of changes to ease review:

1. I’ve split off Development/IDE/Core/RuleTypes/Daml.hs into its own
   Bazel targets to avoid cycles. This allows us to use a rule without
   depending on the implementation of that rule.

2. Development/IDE/Core/IdeState/Daml.hs has moved into its own Bazel
   target. There is a lot of stuff that needs `withDamlIdeState` now
   since the `ghcide` `initialise` function does not add our Shake rule
   for getting the GHC session and making all of that depend on
   damlc:lib creates a mess.

3. The actual logic of the initialization has stayed the same but
   moved into a rule in Development/IDE/Core/Rules/Daml.hs.

4. As mentioned above, a few things that could previously get away
   with only the rules provided by ghcide now need our own rules so
   they switched to `withDamlIdeState`. This mainly affects `damlc
   docs`.

5. Making daml-docs work with the ofInterestRule is a bit tricky.I did
   find the issue and included the fix but also disabled the
   ofInterestRule since it really does not make sense to generate Core
   in the background.

changelog_begin
changelog_end
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs [hunk 1]
```diff
@@ -516,6 +516,25 @@ generatePackageMapRule opts = do
...
+damlGhcSessionRule opts@Options{..} = do
+    defineEarlyCutoff $ \DamlGhcSession projectRoot -> do
+        let base = mkBaseUnits (optUnitId opts)
+        optPackageImports <-
+          if getInferDependantPackages optInferDependantPackages
+          then do
+            more <- liftIO $ dependantUnitsFromDamlYaml optDamlLfVersion (fromNormalizedFilePath projectRoot)
+            pure $ map mkPackageFlag (base ++ more) ++ optPackageImports
+          else
+            pure $ map mkPackageFlag base ++ optPackageImports
```
Comment (associahedron) [inline compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs:529]: suggestion:

```suggestion
        packages <- if getInferDependentPackages optInferDependantPackages
            then fmap (base ++) . liftIO $ dependentUnitsFromDamlYaml optDamlLfVersion (fromNormalizedFilePath projectRoot)
            else pure base   
        let optPackageImports = map mkPackageFlag packages ++ optPackageImports
```

Comment (cocreature) [reply compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs:529]: Sorry this is still a draft PR, mostly just testing CI. A bunch of stuff is going to change and needs cleanup :slightly_smiling_face:  I’ll ping you once it’s ready

Comment (cocreature) [reply compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs]: Addressed is as part of cleanup, thanks for the suggestion!

**Comment by cocreature** [comment]
This is ready now, I’ve added a description of the changes in the PR description and commit message. It’s mostly just reshuffling a lot of code but no actual changes.

**Comment by cocreature** [comment]
> And sorry for jumping the gun on reviewing the draft PR. :-)

Nothing to be sorry for, more reviews are better :slightly_smiling_face:

**Pull request merged**